pdg/src/info.rs

```diff
@@ -67,18 +67,481 @@ fn set_flow_info(g: &mut Graph) {
         });
     }
 }
 
+/// Gathers information from a [`Graph`] (assumed to be acyclic and topologically sorted but not
+/// necessarily connected) for each [`Node`] in it what its chronologically (judged by [`NodeId`])
+/// final descendent is.
+fn get_last_desc(g: &mut Graph) -> HashMap<NodeId, NodeId> {
+    let mut desc_map: HashMap<NodeId, NodeId> =
+        HashMap::from_iter(g.nodes.iter_enumerated().map(|(idx, _)| (idx, idx)));
+    for (n_id, node) in g.nodes.iter_enumerated().rev() {
+        if let Some(p_id) = node.source {
+            let cur_node_last_desc: NodeId = *desc_map.get(&n_id).unwrap();
+            let parent_last_desc: NodeId = desc_map.remove(&p_id).unwrap();
+            desc_map.insert(p_id, std::cmp::max(cur_node_last_desc, parent_last_desc));
+        }
+    }
+    desc_map
+}
+
+/// Finds the inverse of a [`Graph`], each [`Node`] mapping to a list of its children.
+fn collect_children(g: &Graph) -> HashMap<NodeId, Vec<NodeId>> {
+    let mut m = HashMap::new();
+    for (par, chi) in g
+        .nodes
+        .iter_enumerated()
+        .filter_map(|(idx, node)| Some((node.source?, idx)))
+    {
+        m.entry(par).or_insert_with(Vec::new).push(chi)
+    }
+    for par in g.nodes.indices() {
+        let _ = m.try_insert(par, Vec::new());
+    }
+    m
+}
+
+/// Given a list of nodes of the same parent and information about them,
+/// determines if any have conflicts with any of the others.
+/// Children which are not a field cannot be live at the same time as any other child.
+/// Children which are a field cannot be live at the same time as any other one of the same field.
+fn check_children_conflict(
+    g: &Graph,
+    n_id: &NodeId,
+    children: &HashMap<NodeId, Vec<NodeId>>,
+    descs: &HashMap<NodeId, NodeId>,
+) -> bool {
+    let mut max_descs: HashMap<Option<Field>, NodeId> = HashMap::new();
+    for id in children.get(n_id).unwrap() {
+        let sib_node: &Node = g.nodes.get(*id).unwrap();
+        let my_last_desc = descs.get(&id).unwrap().clone();
+        if matches!(max_descs.get(&None), Some(max_desc) if max_desc > id)
+            || matches!(sib_node.kind,NodeKind::Field(f) if matches!(max_descs.get(&Some(f)),Some(max_desc_field) if max_desc_field > id))
+        {
+            return true;
+        }
+        let my_entry: Entry<_, _> = if let NodeKind::Field(f) = sib_node.kind {
+            max_descs.entry(Some(f))
+        } else {
+            max_descs.entry(None)
+        };
+        my_entry
+            .and_modify(|past_last_desc| *past_last_desc = max(*past_last_desc, my_last_desc))
+            .or_insert(my_last_desc);
+    }
+    false
+}
+
+fn set_uniqueness(g: &mut Graph) {
+    let children = collect_children(g);
+    let last_descs = get_last_desc(g);
+    let mut nonuniqueness: HashSet<NodeId> = HashSet::new();
+    for (n_id, node) in g.nodes.iter_enumerated() {
+        if matches!(node.source, Some(p_id) if nonuniqueness.contains(&p_id))
```

Comment:
does this imply that the whole connected tree is non-unique if any nodes contained in it are?

Reply:
There are test cases in which the other direction doesn't hold - if there's a conflict between A's child B and A's child C, A can't be unique, but its parent Z can still be.

Reply:
the reason I asked is because I can't tell whether or not the order of traversal in nodes matters or not, but it sounds like it doesn't because it's strictly limited to immediate parent and children. is that right?

Reply:
Order of traversal does matter, because any descendant of a non-unique node is non-unique. My assumption which I believe is listed in the comments is that the nodes are ordered.